### PR TITLE
ENH: add np.positive ufunc

### PR DESCRIPTION
**Pull request by shoyer**
xref #8932

The strategy here was to search for every use of "negative" in NumPy's C code, so I quite possibly missed something. But it builds, and seems to work.

Still needs full test coverage. The existing test coverage for ufuncs like absolute and negative is rather sparse, but that's no excuse for new code.

Questions to resolve:
- Is there a better way to do this, perhaps still making use of PyArray_Copy?
- Do I need to write an simd loop for floats? This looked a little tricky, so I didn't bother. But it does mean that positive is slightly slower than negative for floats (I measured 7% for a large array), which is a little strange.
- Do we need to deprecate the existing behavior for the unary op instead of suddenly raising for operations like `+np.array('foo')`? How do we even detect these cases? I guess we can check for dtypes that are no longer valid (e.g., strings), but there's no way to know if unary `+` is a valid on an dtype=object array without trying it on each of the scalars.
- Should `+` on a boolean array be valid? I guess the answer is no, given that `-x` raises for boolean arrays and based on the principle that `+x` should only be valid when `-x` is. This strengths the case for a deprecation cycle for the behavior of `ndarray.__pos__`, given that we went through such a deprecation cycle for negative on booleans (https://github.com/numpy/numpy/pull/8297).

### CODE REVIEW

**Comment by shoyer** [comment]
For now, it might be better to skip trying to fix up `ndarray.__pos__`, and only add the new ufunc for use with NDArrayOperatorsMixin.

File: numpy/core/src/umath/loops.c.src [hunk 1]
```diff
@@ -874,6 +874,14 @@ NPY_NO_EXPORT NPY_GCC_OPT_3 @ATTR@ void
 
 #if @CHK@
 NPY_NO_EXPORT NPY_GCC_OPT_3 @ATTR@ void
+@TYPE@_positive@isa@(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
+{
+    UNARY_LOOP_FAST(@type@, @type@, *out = in);
```
Comment (eric-wieser) [inline numpy/core/src/umath/loops.c.src]: Can we do `*out = +in` here, for clarity?

File: numpy/core/src/umath/loops.c.src [hunk 2]
```diff
@@ -2189,6 +2215,15 @@ HALF_negative(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUS
 }
 
 NPY_NO_EXPORT void
+HALF_positive(char **args, npy_intp *dimensions, npy_intp *steps, void *NPY_UNUSED(func))
+{
+    UNARY_LOOP {
+        const npy_half in1 = *(npy_half *)ip1;
+        *((npy_half *)op1) = in1;
+    }
```
Comment (eric-wieser) [inline numpy/core/src/umath/loops.c.src]: I realize that you're just copying `negative` here - but what's the heuristic for when to use `UNARY_LOOP_FAST` and when to just use `UNARY_LOOP`? @juliantaylor, I suspect you can shed some light here?

Comment (juliantaylor) [inline numpy/core/src/umath/loops.c.src]: you use `UNARY_LOOP_FAST` in cases where compiler vectorization is profitable.
It involves several copies of the loop so it is just bloat in cases where no vectorization can be done like in the negative half case.
You could use a fast loop for positive here (though it is hardly worth the extra generated code)
In general you could use a memcpy special case instead of the fast loop here too, like the logical_not in gh-8924. Some compilers might not be able to do that for you.

File: numpy/core/src/multiarray/number.c [hunk 3]
```diff
@@ -1128,7 +1139,7 @@ NPY_NO_EXPORT PyNumberMethods array_as_number = {
     (binaryfunc)array_divmod,                   /*nb_divmod*/
     (ternaryfunc)array_power,                   /*nb_power*/
     (unaryfunc)array_negative,                  /*nb_neg*/
-    (unaryfunc)_array_copy_nice,                /*nb_pos*/
```
Comment (juliantaylor) [inline numpy/core/src/multiarray/number.c]: this function can be removed now

File: numpy/core/src/umath/loops.c.src (hunk 1 above)
Comment (shoyer) [inline numpy/core/src/umath/loops.c.src]: Done.

File: numpy/core/src/multiarray/number.c (hunk 3 above)
Comment (shoyer) [inline numpy/core/src/multiarray/number.c]: With some regret, I decided to undo this change, because of the potential for a backwards compatibility break (we're pretty paranoid about those).

File: numpy/core/src/umath/loops.c.src (hunk 1 above)
Comment (eric-wieser) [inline numpy/core/src/umath/loops.c.src]: Then you haven't pushed!

**Comment by shoyer** [comment]
Any further concerns here?

Note that I rolled back the changes to `ndarray.__pos__`

**Review by eric-wieser — APPROVED** [review]

**Comment by eric-wieser** [comment]
Ah, but needs a release note

**Comment by eric-wieser** [comment]
Proposal - can all the `positive` loops be combined into the same template as the `negative` loops? So add a `repeat` with `@op = +,-@`. Just in the interest of keeping what is an already large file to a somewhat reasonable size.

**Comment by charris** [comment]
@shoyer Needs a rebase and release note.

**Comment by shoyer** [comment]
> Proposal - can all the positive loops be combined into the same template as the negative loops? So add a repeat with @op = +, -@, @func = positive, negative@. Just in the interest of keeping what is an already large file to a somewhat reasonable size.

This is a good idea, but it turned out to be easier said that done:
- For integers, negative uses a non-trivial SIMD trick
- For timedelta, negative needs to check for NaT
- For floats, negative also uses SIMD
- For half, negative also uses something clever

So I just stuck with separate definitions.

> Needs a rebase and release note

Done and done.

**Comment by eric-wieser** [comment]
Thanks @shoyer!

**Comment by mhvk** [comment]
I'm puzzled, this one function does not seem to work well with `__array_ufunc__`:
```
class A(np.ndarray):
    def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
        return "A"

a = np.array(1.).view(A)
-a
# 'A'
+a
# 1.0
```

I think it is due to `positive` still being defined as `_array_copy_nice` in `number.c`, but simply changing that and adding to `GET` and `SET` seemed insufficient:
```
numpy/core/src/multiarray/number.c:573:50: error: ‘NumericOps {aka struct <anonymous>}’ has no member named ‘positive’; did you mean ‘negative’?
     return PyArray_GenericUnaryFunction(m1, n_ops.positive);
```
(will try to investigate later, but have to run now; if someone else understands what is happening, do go ahead and fix...)

**Comment by eric-wieser** [comment]
@shoyer deliberately did not connect up `np.positive` to `__pos__` yet, so I think this is expected behaviour?

**Comment by mhvk** [comment]
OK, rereading the comments I see that is the case. I guess I'll have to keep `Quantity.__pos__` for another release then...

**Comment by mhvk** [comment]
Probably being careful is good, but should we at least add the deprecation warning? Also, maybe just pass on to `np.positive` for classes that define `__array_ufunc__`? (Since those obviously know what is happening.)

**Comment by shoyer** [comment]
@mhvk Those both sound like fine ideas to me.

**Comment by mhvk** [comment]
@shoyer - would you be able to do it? Or have an idea what I should do beyond what I tried (https://github.com/numpy/numpy/pull/8967#issuecomment-299955657)

**Comment by shoyer** [comment]
@mhvk I'm not sure I'll have time to get to this before the branch for 1.13. Probably, the desired behavior is something like:
```
def __pos__(self):
    try:
        return np.positive(self)
    except TypeError:
        issue deprecation warning
        return self.copy()
```
(This may be a little tricky, though, because it will need to be done in C.)

For your ndarray subclass, you could overwrite `__pos__` to call `np.positive`, though would probably need a fallback option for older versions of numpy where `np.positive` is not available.

**Comment by mhvk** [comment]
The override is indeed what I do for `Quantity`; I'm also not sure I'll have time for another crack at this.

**Comment by shoyer** [comment]
I created https://github.com/numpy/numpy/issues/9081 so we don't forget about this.
